codec/src/conformance.rs

```diff
@@ -0,0 +1,452 @@
+    /// Hex-encoded SHA-256 hash of all encoded values concatenated together.
+    pub hash: String,
+}
+
+/// Errors that can occur when loading conformance files.
+#[derive(Debug)]
+pub enum ConformanceError {
+    /// Failed to read the file.
+    Io(std::path::PathBuf, std::io::Error),
+    /// Failed to parse the TOML.
+    Parse(std::path::PathBuf, toml::de::Error),
+}
+
+impl std::fmt::Display for ConformanceError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Io(path, e) => write!(f, "failed to read {}: {}", path.display(), e),
+            Self::Parse(path, e) => write!(f, "failed to parse {}: {}", path.display(), e),
+        }
+    }
+}
+
+impl std::error::Error for ConformanceError {}
+
+impl ConformanceFile {
+    /// Load a conformance file from the given path.
+    pub fn load(path: &Path) -> Result<Self, ConformanceError> {
+        let contents =
+            fs::read_to_string(path).map_err(|e| ConformanceError::Io(path.to_path_buf(), e))?;
+        toml::from_str(&contents).map_err(|e| ConformanceError::Parse(path.to_path_buf(), e))
+    }
+
+    /// Load a conformance file, returning an empty file if it doesn't exist.
+    pub fn load_or_default(path: &Path) -> Result<Self, ConformanceError> {
+        if path.exists() {
+            Self::load(path)
+        } else {
+            Ok(Self::default())
+        }
+    }
+}
+
+/// Encode bytes as a lowercase hex string.
+pub fn hex_encode(bytes: &[u8]) -> String {
+    const HEX_CHARS: &[u8; 16] = b"0123456789abcdef";
+    let mut result = String::with_capacity(bytes.len() * 2);
+    for &byte in bytes {
+        result.push(HEX_CHARS[(byte >> 4) as usize] as char);
+        result.push(HEX_CHARS[(byte & 0x0f) as usize] as char);
+    }
+    result
+}
+
+/// Size of the random buffer used for generating arbitrary values.
+///
+/// This should be large enough to generate complex nested types.
+const ARBITRARY_BUFFER_SIZE: usize = 4096;
+
+/// Generate a deterministic value of type `T` using the given seed.
+///
+/// Uses the `arbitrary` crate with a seeded ChaCha RNG for reproducible
+/// value generation.
+pub fn generate_value<T>(seed: u64) -> T
+where
+    T: for<'a> Arbitrary<'a>,
+{
+    // Create a seeded RNG
+    let mut rng = ChaCha8Rng::seed_from_u64(seed);
+
+    // Generate random bytes for the Unstructured input
+    let mut buffer = vec![0u8; ARBITRARY_BUFFER_SIZE];
+    rng.fill(&mut buffer[..]);
+
+    // Generate the arbitrary value
+    let mut unstructured = Unstructured::new(&buffer);
+    T::arbitrary(&mut unstructured).expect("failed to generate arbitrary value")
+}
+
+/// Compute the conformance hash for a type.
+///
+/// Generates `n_cases` arbitrary values, encodes each one, and hashes
+/// all the encoded bytes together using SHA-256.
+pub fn compute_conformance_hash<T>(n_cases: usize) -> String
+where
+    T: Encode + for<'a> Arbitrary<'a>,
+{
+    let mut hasher = Sha256::new();
```

Comment:
We should use the `Transcript` API instead.

Reply:
See: https://github.com/commonwarexyz/monorepo/issues/2258#issuecomment-3639513300